src/daft-core/src/datatypes/logical.rs

```diff
@@ -1,121 +1,153 @@
 use std::{marker::PhantomData, sync::Arc};
 
-use crate::datatypes::{BooleanArray, DaftLogicalType, DateType, Field};
+use crate::{
+    datatypes::{DaftLogicalType, DateType, Field},
+    with_match_daft_logical_primitive_types,
+};
 use common_error::DaftResult;
 
 use super::{
-    DataArray, DataType, Decimal128Type, DurationType, EmbeddingType, FixedShapeImageType,
-    FixedShapeTensorType, ImageType, TensorType, TimestampType,
+    nested_arrays::FixedSizeListArray, DaftArrayType, DaftDataType, DataArray, DataType,
+    Decimal128Type, DurationType, EmbeddingType, FixedShapeImageType, FixedShapeTensorType,
+    ImageType, TensorType, TimestampType,
 };
-pub struct LogicalArray<L: DaftLogicalType> {
+
+/// A LogicalArray is a wrapper on top of some underlying array, applying the semantic meaning of its
+/// field.datatype() to the underlying array.
+#[derive(Clone)]
+pub struct LogicalArrayImpl<L: DaftLogicalType, PhysicalArray: DaftArrayType> {
     pub field: Arc<Field>,
-    pub physical: DataArray<L::PhysicalType>,
+    pub physical: PhysicalArray,
     marker_: PhantomData<L>,
 }
 
-impl<L: DaftLogicalType> Clone for LogicalArray<L> {
-    fn clone(&self) -> Self {
-        LogicalArray::new(self.field.clone(), self.physical.clone())
-    }
-}
+impl<L: DaftLogicalType, W: DaftArrayType> DaftArrayType for LogicalArrayImpl<L, W> {}
 
-impl<L: DaftLogicalType + 'static> LogicalArray<L> {
-    pub fn new<F: Into<Arc<Field>>>(field: F, physical: DataArray<L::PhysicalType>) -> Self {
+impl<L: DaftLogicalType, P: DaftArrayType> LogicalArrayImpl<L, P> {
+    pub fn new<F: Into<Arc<Field>>>(field: F, physical: P) -> Self {
         let field = field.into();
         assert!(
             field.dtype.is_logical(),
             "Can only construct Logical Arrays on Logical Types, got {}",
             field.dtype
         );
-        assert_eq!(
-            physical.data_type(),
-            &field.dtype.to_physical(),
-            "Expected {} for Physical Array, got {}",
-            &field.dtype.to_physical(),
-            physical.data_type()
-        );
-
-        LogicalArray {
+        // TODO(FixedSizeList): How to do this assert on the physical datatype?
```

Comment:
This assert is no longer possible unless we add `fn datatype` to the `DaftArrayType` trait, or we implement `new` on each concrete type in our macro `impl_logical_type!` below.